subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/DefaultIvyPublication.java

```diff
@@ -245,15 +257,54 @@ private void populateFromComponent() {
                     if (dependency instanceof ProjectDependency) {
                         addProjectDependency((ProjectDependency) dependency, confMapping);
                     } else {
-                        addExternalDependency((ExternalDependency) dependency, confMapping);
+                        ExternalDependency externalDependency = (ExternalDependency) dependency;
+                        if (PlatformSupport.isTargettingPlatform(dependency)) {
+                            if (unsupportedUsages == null) {
+                                unsupportedUsages = Sets.newHashSet();
+                            }
+                            unsupportedUsages.add(String.format("%s:%s:%s declared as platform", dependency.getGroup(), dependency.getName(), dependency.getVersion()));
+                        }
+                        if (externalDependency.getVersion() == null) {
+                            if (unsupportedUsages == null) {
+                                unsupportedUsages = Sets.newHashSet();
+                            }
+                            unsupportedUsages.add(String.format("%s:%s declared without version", externalDependency.getGroup(), externalDependency.getName()));
+                        }
+                        addExternalDependency(externalDependency, confMapping);
                     }
                 }
             }
+            if (!usageContext.getDependencyConstraints().isEmpty()) {
+                if (unsupportedUsages == null) {
+                    unsupportedUsages = Sets.newHashSet();
+                }
+                for (DependencyConstraint constraint : usageContext.getDependencyConstraints()) {
+                    unsupportedUsages.add(String.format("%s:%s:%s declared as a dependency constraint", constraint.getGroup(), constraint.getName(), constraint.getVersion()));
+                }
+            }
+            if (!usageContext.getCapabilities().isEmpty()) {
+                if (unsupportedUsages == null) {
+                    unsupportedUsages = Sets.newHashSet();
+                }
+                for (Capability capability : usageContext.getCapabilities()) {
+                    unsupportedUsages.add(String.format("Declares capability %s:%s:%s", capability.getGroup(), capability.getName(), capability.getVersion()));
+                }
+            }
 
             for (ExcludeRule excludeRule : usageContext.getGlobalExcludes()) {
                 globalExcludes.add(new DefaultIvyExcludeRule(excludeRule, conf));
             }
         }
+        if (unsupportedUsages != null) {
+            TreeFormatter formatter = new TreeFormatter();
+            formatter.node(getDisplayName() + INCOMPATIBLE_FEATURE);
+            formatter.startChildren();
+            for (String usage : unsupportedUsages) {
+                formatter.node(usage);
+            }
+            formatter.endChildren();
+            LOG.lifecycle(formatter.toString());
+        }
```

Comment:
Most of this validation code can be extracted into a utility that is shared between Maven and Ivy publishing. This might be a simple builder that allows you to add an 'unsupportedUsage', and it constructs the message.

Reply:
Done